### PR TITLE
[API][Queues] [SIMS #1449]Send student notifications

### PR DESCRIPTION
**Pull request by dheepak-aot**
## Send student notifications
Added student notifications for following scenarios. Creating a scheduler to send the notification by calling GC notify will be in next PR after when https://github.com/bcgov/SIMS/pull/1604 is merged.

- [x] Scenario - Ministry completes or denies an exception
- Template name: Student - Student - Ministry Completes Exception
- givenNames, lastName, email address, date
- message sentence: Your exception request has been completed. Please visit your application details page to view additional information.

- [x] Scenario - Ministry completes or denies a change request
- Template name: Student - Student - Ministry Completes Requested Change
- givenNames, lastName, email address, date
- message sentence: Your requested change has been completed. Please visit your application details page to view additional information.

- [x] Scenario - School reports a change
- Template name: Student - Student - Institution Reports A Change
- givenNames, lastName, email address, date
- message sentence: Your institution has reported a change to your application. Please visit your application details page to view additional information.

- [x] Scenario - School completes a PIR
- Template name: Student - Student - Institution Completes Program Information Request
- givenNames, lastName, email address, date
- message sentence: Your institution has completed your program information request. Please visit your application details page to view additional information.

- [x] Scenario - School confirms Enrollment (including tuition remittance status)
- Template name: Student - Student - Institution Confirms Enrolment
- givenNames, lastName, email address, date
- message sentence: Your institution has completed your confirmation of enrolment. Please visit your application details page to view additional information.

### CODE REVIEW

File: sources/packages/backend/apps/api/src/services/student-file/student-file.service.ts [hunk 1]
```diff
@@ -113,17 +112,15 @@ export class StudentFileService extends RecordDataModelService<StudentFile> {
    * @param uniqueFileNames list of unique file names.
    * @param fileOrigin origin of the file being saved.
    * @param groupName group name of the file being save.
-   * @param sendNotification optional notification message to be sent.
-   * @param metadata optional metadata of the file being save.
+   * @param options file upload options.
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/services/student-file/student-file.service.ts:115]: When we move the option to an object I would recommend adding the comments there.
If we create the inline object we should also add the comments here. Either way, the comments are welcome.

File: sources/packages/backend/apps/api/src/esdc-integration/fed-restriction-integration/fed-restriction-processing.service.ts [hunk 2]
```diff
@@ -262,7 +262,6 @@ export class FedRestrictionProcessingService {
           await this.studentRestrictionService.createNotifications(
             insertedRestrictionsIDs,
             auditUserId,
-            { notificationsDelayed: true },
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/esdc-integration/fed-restriction-integration/fed-restriction-processing.service.ts:265]: why was it removed

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 3]
```diff
@@ -118,11 +113,11 @@ export class NotificationActionsService {
    * @param auditUserId user that should be considered the one that is causing the changes.
    * @param options options for the student restriction notification.
    */
-  async sendStudentRestrictionAddedNotification(
+  async saveStudentRestrictionAddedNotification(
     notifications: StudentRestrictionAddedNotification[],
     auditUserId: number,
     options?: StudentRestrictionAddedNotificationOptions,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:119]: I believe that we can get rid of this class now and make this method the same as the others.

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 4]
```diff
@@ -118,11 +113,11 @@ export class NotificationActionsService {
...
-  ): Promise<void> {
+  ): Promise<number[]> {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:120]: Please update @return or change it to void.

File: sources/packages/backend/libs/sims-db/src/entities/notification.model.ts [hunk 5]
```diff
@@ -86,4 +86,29 @@ export enum NotificationMessageType {
    * 3. Federal restriction placed on student account.
    */
   StudentRestrictionAdded = 3,
+
+  /**
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/sims-db/src/entities/notification.model.ts:90]: Can we please keep the file consistent and remove the blank likes?

**Review by andrewsignori-aot — COMMENTED** [review]
Great work, please take a look at the comments.

File: sources/packages/backend/apps/api/src/services/application-exception/application-exception.service.ts [hunk 6]
```diff
@@ -155,7 +164,21 @@ export class ApplicationExceptionService extends RecordDataModelService<Applicat
...
       exceptionToUpdate.modifier = auditUser;
       exceptionToUpdate.updatedAt = now;
-      return applicationExceptionRepo.save(exceptionToUpdate);
+
+      const exception = await applicationExceptionRepo.save(exceptionToUpdate);
+
+      // Create a student notification when ministry completes an exception.
+      await this.notificationActionsService.saveExceptionCompleteNotification(
+        {
+          givenNames: exceptionToUpdate.application.student.user.firstName,
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/services/application-exception/application-exception.service.ts:173]: just a suggestion, as we did in other parts, may u can assign a user to a const and use it. 
```
const user = exception.application.student.user;

lastName: user.lastName,
 toAddress: user.email,
 userId: user.id,
```

File: sources/packages/backend/apps/api/src/services/application/application.service.ts [hunk 7]
```diff
@@ -736,7 +738,22 @@ export class ApplicationService extends RecordDataModelService<Application> {
         auditUserId,
         transactionalEntityManager,
       );
-      return applicationRepo.save(application);
+
+      const updatedApplication = await applicationRepo.save(application);
+
+      //Create student notification when institution completes PIR.
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/services/application/application.service.ts:744]: pls add a space after `//`

File: sources/packages/backend/apps/api/src/services/application/application.service.ts [hunk 8]
```diff
@@ -983,40 +1000,71 @@ export class ApplicationService extends RecordDataModelService<Application> {
...
+        );
+      }
+      application.pirDeniedOtherDesc = otherReasonDesc;
+      application.pirStatus = ProgramInfoStatus.declined;
+      application.modifier = { id: auditUserId } as User;
+      await transactionalEntityManager
+        .getRepository(Application)
+        .save(application);
+
+      //Create student notification when institution completes PIR.
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/services/application/application.service.ts:1055]: same here and for others

File: sources/packages/backend/apps/api/src/services/disbursement-schedule-service/disbursement-schedule-service.ts [hunk 9]
```diff
@@ -612,4 +628,53 @@ export class DisbursementScheduleService extends RecordDataModelService<Disburse
       })
       .getMany();
   }
+
+  /**
+   * Create institution confirm COE notification to notify student
+   * when institution confirms a COE to their application.
+   * @param disbursementScheduleId
+   * @param auditUserId
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/services/disbursement-schedule-service/disbursement-schedule-service.ts:636]: Pls add brief comments to the params

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 10]
```diff
@@ -62,25 +63,22 @@ export class NotificationActionsService {
...
    * to execute the command as part of an existing transaction. If not provided
    * the local repository will be used instead.
    */
-  async sendMinistryFileUploadNotification(
+  async saveMinistryFileUploadNotification(
     notification: MinistryStudentFileUploadNotification,
     auditUserId: number,
     entityManager?: EntityManager,
-  ): Promise<void> {
+  ): Promise<number> {
```
Comment (ann-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:81]: pls add the comment for the return

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts (hunk 4 above)
Comment (ann-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:120]: pls add the comment for the return

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 11]
```diff
@@ -142,33 +137,211 @@ export class NotificationActionsService {
...
+   * @param notification notification details.
+   * @param auditUserId user who reports the change.
+   * @param entityManager entity manager to execute in transaction.
+   * @returns notification created.
+   */
+  async saveInstitutionReportChangeNotification(
+    notification: StudentNotification,
+    auditUserId: number,
+    entityManager: EntityManager,
+  ) {
```
Comment (ann-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:239]: pls define the return `: Promise<number>`

File: sources/packages/backend/apps/api/src/services/student-file/student-file.model.ts [hunk 12]
```diff
@@ -1,7 +1,21 @@
...
   fileName: string;
   uniqueFileName: string;
   groupName: string;
   mimeType: string;
   fileContent: Buffer;
 }
+
+export interface FileUploadOptions {
+  /**
+   * optional notification message to be sent
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/services/student-file/student-file.model.ts:14]: Upper case and period 😉

File: sources/packages/backend/apps/api/src/services/student-file/student-file.model.ts [hunk 13]
```diff
@@ -1,7 +1,21 @@
...
+   */
+  saveFileUploadNotification?: (entityManager: EntityManager) => Promise<void>;
+  /**
+   * optional metadata of the file being saved.
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/api/src/services/student-file/student-file.model.ts:18]: Upper case.

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 14]
```diff
@@ -100,28 +95,24 @@ export class NotificationActionsService {
...
-    );
   }
 
   /**
    * Creates a new notification when a new restriction is added to the student account.
    * @param notifications notifications information.
    * @param auditUserId user that should be considered the one that is causing the changes.
-   * @param options options for the student restriction notification.
+   * @param entityManager entity manager to execute in transaction.
+   * @returns notification id created.
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:110]: Please remove the `@returns`.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for doing the changes. Added some minor comments but they are not blockers, hence approving it.

File: sources/packages/backend/apps/api/src/services/student-file/student-file.service.ts (hunk 1 above)
Comment (ann-aot) [reply sources/packages/backend/apps/api/src/services/student-file/student-file.service.ts:115]: did you add it? I am not seeing it

File: devops/openshift/queue-consumers-deploy.yml [hunk 15]
```diff
@@ -120,6 +120,21 @@ objects:
...
+                - name: GC_NOTIFY_API_KEY
+                  valueFrom:
+                    secretKeyRef:
+                      key: ${GC_NOTIFY_API_KEY_NAME_KEY}
+                      name: ${QUEUE_CONSUMERS_SECRET_NAME}
+                - name: GC_NOTIFY_TO_ADDRESS
+                  valueFrom:
+                    secretKeyRef:
+                      key: ${GC_NOTIFY_TO_ADDRESS_NAME_KEY}
+                      name: ${QUEUE_CONSUMERS_SECRET_NAME}
```
Comment (guru-aot) [inline devops/openshift/queue-consumers-deploy.yml:137]: As the queue secrets are growing and is not shared, lets discuss if we need separate secrets for queues during our dev sync up.

File: sources/packages/backend/apps/api/src/services/student-file/student-file.service.ts (hunk 1 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/services/student-file/student-file.service.ts:115]: Added to the properties in options object.
image

File: sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts [hunk 16]
```diff
@@ -357,9 +364,22 @@ export class StudentAppealService extends RecordDataModelService<StudentAppeal>
...
 
-      return transactionalEntityManager
+      const updatedStudentAppeal = await transactionalEntityManager
         .getRepository(StudentAppeal)
         .save(appealToUpdate);
+
+      // Create student notification when ministry completes student appeal.
+      await this.notificationActionsService.saveChangeRequestCompleteNotification(
+        {
+          givenNames: appealToUpdate.application.student.user.firstName,
```
Comment (ann-aot) [inline sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts:374]: u missed this

**Review by ann-aot — APPROVED** [review]
Good work @dheepak-aot . minor comments

File: sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts (hunk 16 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/api/src/services/student-appeal/student-appeal.service.ts]: Updated.

**Pull request merged**